torch_geometric/data/hetero_data.py

```diff
@@ -778,6 +779,71 @@ def edge_type_subgraph(self, edge_types: List[EdgeType]) -> Self:
                 del data[node_type]
         return data
 
+    def concat(self, hetero_data: Self) -> Self:
+        r"""Concatenates :obj:`self` with another :obj:`hetero_data` object.
+        All values needs to have matching shapes at non-concat dimensions.
+        """
+        out = copy.copy(self)
+        for store, other_store in zip(out.stores, hetero_data.stores):
+            store.concat(other_store)
+        return out
+
+    def is_sorted_by_time(
+            self, types: Union[List[EdgeType], List[NodeType]]) -> bool:
+        r"""Returns :obj:`True` if :obj:`time` exists and is sorted for all
+        :obj:`types` (either edge types or node types).
+        """
+        if not types:
+            return False
+
+        for t in types:
+            if 'time' not in self[t]:
+                return False
+            if not self[t].is_sorted_by_time():
+                return False
+
+        return True
+
+    def sort_by_time(self, types: Union[List[EdgeType],
+                                        List[NodeType]]) -> Self:
+        r"""Sorts data associated with :obj:`time` according to :obj:`time` for
+        all :obj:`types` (either edge types or node types).
+        """
+        if not types:
+            return self
+
+        out = copy.copy(self)
+        for t in types:
+            out[t].sort_by_time()
+        return out
+
+    def snapshot(self, types: Union[List[EdgeType], List[NodeType]],
```

Comment:
I have a question about `snapshot` which is relevant to this PR and #8454. 
Assume nodes n1 and n2 with time equals 1 and 4 respectively, and an edge e1 (between n1 and n2 ) with time = 5. 
On calling `snapshot(start_time = 4, end_time = 5)` what would the output look like, from the code it looks like the returned graph will have node n2 and edge e1, but e1 doens't have a source node.

Reply:
In the current design, `time` cannot be an edge and node attribute at the same time, so this scenario is impossible. As these functionalities are moved directly from `TemporalData` I would consider underlying data rather as a list of events.

Reply:
Is there some part of the code that explicitly prevents this?
Does this also apply to hetero graphs?

Reply:
2x Yes, for example at [BaseStorage](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/data/storage.py#L360) and [NeighborSampler](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/sampler/neighbor_sampler.py#L75).